server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -1304,97 +1316,89 @@ private static IndexMetadata.State excludeState(IndicesOptions options) {
             return excludeState;
         }
 
-        private static Map<String, IndexAbstraction> matches(Context context, String expression) {
-            SortedMap<String, IndexAbstraction> indicesLookup = context.getState().getMetadata().getIndicesLookup();
-            if (Regex.isMatchAllPattern(expression)) {
-                return filterIndicesLookup(indicesLookup, null, context.getOptions().ignoreAliases(), context.includeDataStreams());
-            } else if (expression.indexOf("*") == expression.length() - 1) {
-                return suffixWildcard(indicesLookup, expression, context.getOptions().ignoreAliases(), context.includeDataStreams());
+        /**
+         * Given a single wildcard {@param expression}, return the {@code Stream} that contains all the resources (i.e. indices, aliases,
+         * and datastreams), that exist in the cluster at this moment in time, and that the wildcard "resolves" to (i.e. the resource's
+         * name matches the {@param expression} wildcard).
+         * The {@param context} provides the current time-snapshot view of all the resources, as well as conditions
+         * on whether to consider alias, datastream, system, and hidden resources.
+         * It does NOT consider the open or closed status of index resources.
+         */
+        private static Stream<IndexAbstraction> matchResourcesToWildcard(Context context, String wildcardExpression) {
+            assert Regex.isSimpleMatchPattern(wildcardExpression);
+            final SortedMap<String, IndexAbstraction> indicesLookup = context.getState().getMetadata().getIndicesLookup();
+            Stream<IndexAbstraction> matchesStream;
+            if (Regex.isSuffixMatchPattern(wildcardExpression)) {
+                // this is an initial pre-filtering in the case where the expression is a common suffix wildcard, eg "test*"
+                matchesStream = filterIndicesLookupForSuffixWildcard(indicesLookup, wildcardExpression).values().stream();
             } else {
-                return filterIndicesLookup(
-                    indicesLookup,
-                    e -> Regex.simpleMatch(expression, e.getKey()),
-                    context.getOptions().ignoreAliases(),
-                    context.includeDataStreams()
-                );
+                matchesStream = indicesLookup.values().stream();
+            }
+            if (context.getOptions().ignoreAliases()) {
+                matchesStream = matchesStream.filter(indexAbstraction -> indexAbstraction.getType() != Type.ALIAS);
+            }
+            if (Regex.isMatchAllPattern(wildcardExpression) == false && Regex.isSuffixMatchPattern(wildcardExpression) == false) {
```

Comment:
Is there any reason to filter by type before we filter by the expression? I think the code might be clearer if we moved this conditional in the else statement of `if (Regex.isSuffixMatch...`

Reply:
I also think it would be clearer. I kept the original order, though.
Any thoughts @jakelandis ?

Reply:
I've pushed [1b065c8](https://github.com/elastic/elasticsearch/pull/89622/commits/1b065c8350657f7fb0a91bf8f1e657e4f7aa5fe3) . The order can theoretically influence performance, because the alias filter is cheaper from the regex, but it's hard to quantify.